### PR TITLE
whisper: final refactoring

### PR DESCRIPTION
**Pull request by gluk256**
- fixed error caused by the use of uninitialized variable shh
- fixed message saving (envelopes instead of messages)
- fixed filereader mode
- fixed user interface

### CODE REVIEW

File: cmd/wnode/main.go [hunk 1]
```diff
@@ -306,7 +309,9 @@ func startServer() error {
 		configureNode()
 	}
 
-	if !*forwarderMode {
+	if *fileExMode || *fileReader {
+		fmt.Printf("Please type the file name. To quit type: '%s'\n", quitCommand)
```
Comment (gballet) [inline cmd/wnode/main.go]: the file name to do what? You should write the file name to save your messages.

File: cmd/wnode/main.go [hunk 2]
```diff
@@ -197,6 +197,8 @@ func initialize() {
 		if len(*argIP) == 0 {
 			argIP = scanLineA("Please enter your IP and port (e.g. 127.0.0.1:30348): ")
 		}
+	} else if *fileReader {
+		*bootstrapMode = true
```
Comment (gballet) [inline cmd/wnode/main.go]: why would I need to force bootstrap mode here? if one server is a file sharing service available to all, its clients might not want to create their own network

File: cmd/wnode/main.go [hunk 3]
```diff
@@ -643,22 +655,24 @@ func writeMessageToFile(dir string, msg *whisper.ReceivedMessage) {
...
-		} else {
-			fmt.Printf("\n%s {%x}: message received and saved as '%s' (%d bytes)\n", timestamp, address, name, len(msg.Raw))
-		}
-	} else {
-		fmt.Printf("\n%s {%x}: message received (%d bytes), but not saved: %s\n", timestamp, address, len(msg.Raw), name)
+	fullpath := filepath.Join(dir, name)
+	err := ioutil.WriteFile(fullpath, env.Data, 0644)
+	if err != nil {
+		fmt.Printf("\n%s {%x}: message received but not saved: %s\n", timestamp, address, err)
+	} else if show {
```
[the remark this answers is not in the document]
Comment (gluk256) [reply cmd/wnode/main.go]: this is not just a log. this is actually user interface. this is what wnode is for. and now, after conducting some tests i realized that it shows the same message twice (in some cases). so, it is necessary to fix. at the moment, i don't see any other way to do it.

File: whisper/whisperv6/envelope.go [hunk 4]
```diff
@@ -269,3 +269,9 @@ func TopicToBloom(topic TopicType) []byte {
 	}
 	return b
 }
+
+func (w *Whisper) GetEnvelope(hash common.Hash) *Envelope {
```
[the remark this answers is not in the document]
Comment (gluk256) [reply whisper/whisperv6/envelope.go]: thanx, fixed

File: cmd/wnode/main.go [hunk 5]
```diff
@@ -682,18 +696,23 @@ func requestExpiredMessagesLoop() {
 	for {
 		timeLow = scanUint("Please enter the lower limit of the time range (unix timestamp): ")
 		timeUpp = scanUint("Please enter the upper limit of the time range (unix timestamp): ")
-		t = scanLine("Please enter the topic (hexadecimal): ")
-		if len(t) >= whisper.TopicLength*2 {
+		t = scanLine("Please enter the topic (hexadecimal). For all topics just press Enter: ")
```
[the remark this answers is not in the document]
Comment (gluk256) [reply cmd/wnode/main.go]: thanx

File: cmd/wnode/main.go (hunk 2 above)
Comment (gluk256) [reply cmd/wnode/main.go]: filereader does not send or receive messages. it only loads encrypted stuff from disc and decrypts. this mode is necessary for diagnostics.

**Pull request merged**